### ISSUE TITLE
Separate convolution from input format

### ISSUE THREAD

**Issue by tyarkoni** [nistats/nistats#181]
If I understand it correctly, currently in `make_design_matrix`, event information passed in as `paradigm` *will* be convolved with the HRF, and values passed to `add_reg` will *not* be convolved. Conceptually, it seems like there should be a decoupling of format from convolution: a user might have good reason to pass values in in either sparse (condition/onset/duration/amplitude) or dense formats (one value per TR), while still wanting to convolve some columns but not others. @bthirion, how would you feel about refactoring the API so that there's something like a `convolve` argument that specifies, by name, which variables to convolve?

To expand, the reason this is an issue for us is that application of transformations in pybids often result in sparse variables being converted to a dense representation. If we feed them to nistats in dense form, they don't get convolved. But if we feed them in as sparse (basically by faking condition/onset/duration information, even though the sampling is uniform), then I assume nistats is going to resample the events a second time, which seems like it introduce additional noise (or maybe not--I haven't checked the implementation).

Although it's not a problem for us, I can also see the opposite case arising--i.e., where somebody has sparse events they *don't* want to convolve with the HRF. In that case, it would be nice if they could just indicate they don't want convolution to be applied, instead of having to densify the events themselves just to avoid convolution.

What do you think?

**Comment by bthirion** [comment]
Let me frame how I understand things:
you currently have 5 time courses of length T=128, e.g.
"""
T = 128
frame_times = np.arange(T)
x = np.random.randn(T, 5)
"""
they correspond to five type of events.
Then you would like to make a design matrix out of them.
What you can do is indeed to sparsify these time courses, creatinga  valid paradigm object:
"""
paradigm = pd.DataFrame({
        'trial_type': np.repeat(np.arange(5), T).astype('str'),
        'onset': np.tile(frame_times, 5),
        'modulation': np.concatenate(x)})
"""
 from which you can create a design matrix
"""
dmtx = make_design_matrix(frame_times, paradigm, drift_model=None)
"""
you mean that this manipulation in too counterintuitive ? I agree it is a bit.

My view so far that additional regressors would be only objects such as motion parameters confound regressors, or sinusoidal regressors used in some paradigms (hrfs) that you don't want to convolve a priori. 
On the other hand, I view what you rightly call the sparse format as a BIDS convention, hence I don't want to go against it...
Hence I'm wondering whether the added complexity makes it worth ?

Regarding no-convolution, you can always use `hrf_model=None`.
WDYT ?

**Comment by tyarkoni** [comment]
Yes, right now we're doing exactly what you do in your example--i.e., basically "faking" the sparse format for dense columns.

Re: setting `hrf_model=None`, I don't think that can be set on a column-by-column basis, so it wouldn't help in the case where a user wants to convolve some columns but not others. Of course one can always "densify" the sparse variable and pass it in as an additional regressor, so with some extra work, the current setup can be made to work for any situation.

I definitely don't think this should be a high priority, but I do think it would be nice to separate the convolution decision from the input format at some point down the line. Right now we're able to make nistats work well with pybids and fitlins, but there's a little bit of extra cognitive load and work involved in figuring out which variables to pass to nistats where. Maybe leave the issue open for now, and hopefully someone in either of our groups can tackle it at some point?

**Comment by bthirion** [comment]
Sure, fair enough.
